### PR TITLE
add  new FW StateIO

### PR DESCRIPTION
**Pull request by moinmoin-sh**

### CODE REVIEW

File: docs/docs/firmware/state-io-esp32.md [hunk 1]
```diff
@@ -2,229 +2,63 @@
...
+This FW combines the functionality of [OXRS-SHA-StateMonitor-ESP32](/docs/firmware/state-monitor-esp32.md) and [OXRS-SHA-StateController-ESP32](/docs/firmware/state-controller-esp32.md).
 
-Cupim drumstick swine venison short loin capicola bacon picanha spare ribs tongue burgdoggen pork chop cow ham. Shank brisket ball tip tri-tip cow jerky shoulder chicken, corned beef alcatra filet mignon. Biltong jerky picanha ham hock swine. Ham hock shank corned beef, porchetta strip steak venison sirloin turducken hamburger ham rump jerky. Frankfurter pork tri-tip beef ribs meatloaf picanha pastrami doner burgdoggen.
+To perform this combined functionality the available number of 128 I/O lines is split into an input partion and an output partition. The split can be configured at runtime.
```
Comment (sumnerboy12) [inline docs/docs/firmware/state-io-esp32.md:12]: Typo `partion` - would suggest `is split into input and output partitions.`

File: docs/docs/firmware/state-io-esp32.md [hunk 2]
```diff
@@ -2,229 +2,63 @@
...
 
 ---
 
 ### How does it work?
-Shoulder picanha tongue cow kevin. Pig alcatra chicken jowl, turducken flank buffalo bresaola burgdoggen ground round venison shoulder pancetta. Sirloin chislic beef ribs turkey biltong. Ground round fatback frankfurter meatball, venison jowl capicola ham hock meatloaf. 
+The maximum number of 128 I/O lines can be achieved by connecting up to 8 I/O-expander chips (MCP23017 with 16 I/Os each) via the I2C bus to the Rack32 Controller. The FW refers to these I/Os with Index numbers (1 .. 128). In the standard configuration all 128 I/O lines are configured as input (StateMonitor) or output (StateController). This FW allows to split the 128 I/O lines into a lower partition of inputs and an upper partition of outputs. With this configurations a single Rack32 Controller can control inputs as well as outputs which could be usefull for small scale or distributed applications. 5 split schemas (Config Options) with different partion sizes are supported.
```
Comment (sumnerboy12) [inline docs/docs/firmware/state-io-esp32.md:17]:
```
The maximum number of 128x I/O lines can be achieved by connecting up to 8x I/O expander chips (MCP23017 with 16x I/Os each) via the I2C bus to the Rack32 controller. The FW refers to these I/O lines with index numbers (1 .. 128). In the standard configuration all 128x I/O lines are configured as input (StateMonitor) or output (StateController). This FW allows you to split the 128x I/O lines into a lower partition of inputs and an upper partition of outputs. With this configurations a single Rack32 controller can monitor inputs as well as control outputs which could be useful for small scale or distributed applications. 

Currently there are 5 supported configurations, with different partition sizes.
```

File: docs/docs/firmware/state-io-esp32.md [hunk 3]
```diff
@@ -2,229 +2,63 @@
...
-Buffalo boudin meatball ball tip spare ribs. Venison cow pig shank pastrami, chislic ham. Beef ribs hamburger sirloin brisket, capicola drumstick cupim sausage shank burgdoggen swine.
-
-| Col 1         | Col 2        | Col 3      |
-| :------------ |:-------------| :--------- |
-| Row data     | Row data      |  Row data  |
-| Row data     | Row data      |  Row data  |
-| Row data     | Row data      |  Row data  |
-
-Sirloin chislic beef ribs turkey biltong. Ground round fatback frankfurter meatball, venison jowl capicola ham hock meatloaf. 
+In the example above the FW is configured for 32 inputs and 96 outputs. The user has to make sure that the connected MCPs are configured properly (correct MCP address). Not all output MCPs need to be populated. The FW is checking the for installed MCPs automatically and allows configuration (indexes) for existing MCPs only
```
Comment (sumnerboy12) [inline docs/docs/firmware/state-io-esp32.md:24]:
```
In the example above the FW is configured for 32x inputs and 96x outputs. The user has to make sure that the connected MCPs are configured properly (correct MCP address). Not all output MCPs need to be populated. The FW will automatically scan for installed MCPs and only allows configuration of existing MCPs.
```

File: docs/docs/firmware/state-io-esp32.md [hunk 4]
```diff
@@ -2,229 +2,63 @@
...
-
-Cupim drumstick swine venison short loin capicola bacon picanha spare ribs tongue burgdoggen pork chop cow ham.
-
-```c
-void setup() {
-  // put your setup code here, to run once:
-  Serial.begin(9600);
-  Serial.println("Hello World!");
-}
+This FW is fully compatible with [OXRS](https://oxrs.io) eco-system and is built based on it's   [Libraries](/docs/libraries/README.md)
```
Comment (sumnerboy12) [inline docs/docs/firmware/state-io-esp32.md:27]: `fully compatible with the [OXRS]...`
`and is built using the standard OXRS [libraries]`

File: docs/docs/firmware/state-io-esp32.md [hunk 5]
```diff
@@ -2,229 +2,63 @@
...
 
-void loop() {
-  // put your main code here, to run repeatedly:
+All the functionality of the [StateMonitor](/docs/firmware/state-monitor-esp32.md) applies to the input partition and [StateController's](/docs/firmware/state-controller-esp32.md) functionality applies to the output partition. Please see there for further details.
```
Comment (sumnerboy12) [inline docs/docs/firmware/state-io-esp32.md:29]: `and the [StateController] functionality`
`Please see their documentation for further details.`

**Review by sumnerboy12 — CHANGES_REQUESTED** [review]
Few minor grammatical suggestions but it looks good.

**Comment by moinmoin-sh** [comment]
I think I followed your suggestions.
Thanks for taking the time and thoroughly reading.

**Pull request merged**